core/builder.go

```diff
@@ -15,6 +15,17 @@ import (
 	"go.uber.org/fx"
 )
 
+type fxOptFunc func([]fx.Option) ([]fx.Option, error)
+
+var fxOptionFuncs []fxOptFunc
+
+// RegisterFXOptionFunc registers a function that is run before the fx app is initialized,
+// with the list of all fx options that will be passed during initialization, and that
+// returns a new list of all fx options.
+func RegisterFXOptionFunc(optFunc fxOptFunc) {
+	fxOptionFuncs = append(fxOptionFuncs, optFunc)
+}
```

Comment:
I get that we're modifying globals to work with the plugin system as we do in some other places. However, I wonder if they shouldn't get passed into `NewNode` (in the build config or as optional arguments)?

It's likely not that big a deal, but I'm mostly worried about if people use kubo as a library rather than using the constituent libraries and start messing with this function. We could either put big warning flags on it, or we could move the singleton pieces closer to the where the executable creates the `NewNode`.

Reply:
Do you have any examples of bad things people would do with this when using kubo as a library?

Another option is to unexport the register func, and pass it as a func into the PluginLoader.

Overall I'm biasing towards keeping it consistent w/ the rest of the plugins so the mechanics and caveats for consumers are the same.

Reply:
> Do you have any examples of bad things people would do with this when using kubo as a library?

Basically if they try and instantiate two kubo nodes in the same process and get confused by global variables. The kubo as a library examples do this, they obviously don't use any custom fx options but if someone tried to fork the example and pass in a custom option things could get confusing.

If you leave it like this I'd add some comments to the function explaining the situation around multiple instantiations, even if it is ultimately just what happens when you use globals for things. Same as the problems with https://github.com/ipfs/go-unixfs/blob/4571136c15d0678006f09ebde615d0c45c126ff9/io/directory.go#L27.

kubo calls `NewNode` a few places within the codebase. Have you taken a look at them to see if it's a good idea for the custom plugins to be shared across all of those places?

> Overall I'm biasing towards keeping it consistent w/ the rest of the plugins so the mechanics and caveats for consumers are the same.

The main reason I'm flagging here to is that this plugin is a little different from the more common usages of the Datastore and IPLD Codec plugins which just add functionality the user may take advantage of. This one can change existing functionality as well.

Reply:
Thanks for the extra info.

Yes I've taken a look at the callers of `NewNode`. There's no real way to know whether it makes sense to apply these custom fx opts to all the various places that `NewNode` is called, it depends on what the options are doing. So I think we should just apply them everywhere and tell plugin authors "your options will be applied every time a node is created", and so it's up to the plugin author to make sure that works in all contexts. There currently isn't a way for a plugin to know what "context" it's in (e.g. "am I in offline mode?")...we could pass that in to the plugin as well, something like:

```go
type NodeInfo {
	FXOpts []fx.Option
	BuildCfg *core.BuildConfig
}
FXOptions(info NodeInfo) ([]fx.Option, error)`)
```

Then the plugin could use that extra info to decide what fx opts to add. (Also, passing a struct here so we can add more fields in the future without breaking the interface).

Would that make sense?

Reply:
Follow up from in person conversation:

> Would that make sense?

It could. We could also just document what's going on like you said and wait on feedback from the first few groups that try using this.

For example, we've seen a few recent cases of people wanting to build plugins we could point this way if/until something better comes along
1. Adding filtering to Bitswap server (who I'll serve data to)
2. Adding filtering to the blockservice for which blocks we'll download
3. Having a custom blockstore that fetches data from a network backend and only processes data with certain codecs

Some of these (e.g. adding allowlists in option 2) will break some node instantiations (e.g. for the node we spin up to do migrations) are potentially problematic.

A reasonable approach is probably to start with globally applied options here with some comments documenting the hazards and then we can add more data to the `NodeInfo` (better name pending) struct to help users disambiguate between what the nodes are used for.

Reply:
Sounds good, I've added these caveates and warnings to the docs, and also switch the input to be a struct so that we can add more info later if we need to.